### ISSUE TITLE
Add metadata to migrations table

### ISSUE THREAD

**Issue by jeremija** [typeorm/typeorm#3375]
**Issue type:**

[x] question
[ ] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:** any

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

I would like to add some metadata to the `migrations` table, like `date_executed` and a `md5sum` or `sha1sum` of the filename to ensure nobody modified a past migration that was already executed. Ideally, it should be possible to add any kind of metadata.

Looks like some methods in `MigrationExecutor` would need to be modified to allow for these changes. If there is interest I could work on a pull request for this.

**Comment by pleerock** [comment]
migrations table is used internally by orm, and if orm doesn't need those additional columns it shouldn't have them. I suggest you to create separate table where you can store what you want.

**Comment by jeremija** [comment]
True, I could create a wrapper around the `migration:run` and `migration:revert` and store this  metadata in a separate table.

However, I just wanted to point out the benefits of the aforementioned columns. A `date_executed` timestamp has proven to be useful when debugging a production instance, and the migration checksum is used to prevent a scenario where new migrations are tested on an empty database (in a CI job), and then a migration suddenly fails on a test or production environment because somebody has modified a past migration. Nothing is worse than having a migration that fails mid-way. Unfortunately MySQL does not really support [transactions for DDL statements](https://dev.mysql.com/doc/refman/8.0/en/cannot-roll-back.html).

Hashing a migration file is new concept:

- [Flyway](https://flywaydb.org/getstarted/how) supports it out of the box (column is called `checksum`)
- It was described in [Stack Overflow: How we do deployment](https://nickcraver.com/blog/2016/05/03/stack-overflow-how-we-do-deployment-2016-edition/#database-migrations) blog post:

   > In memory, we compare the list of migrations that already ran to those needing to run then execute what needs running, in file order. If we find the hash of a filename doesn’t match the migration with the same file name in the table, we abort as a safety measure. We can --force to resolve this in the rare cases a migration should have changed (almost always due to developer error). After all migrations have run, we’re done.

- [postgres-migrations](https://www.npmjs.com/package/postgres-migrations) provides hash checks migration files

   > **Hash checks for previous migrations**
   >
   > Previously run migration scripts shouldn't be modified, since we want the process to be repeated in the same way for every new environment.
   >
   > This is enforced by hashing the file contents of a migration script and storing this in migrations table. Before running a migration, the previously run scripts are hashed and checked against the database to ensure they haven't changed.

- [sqlmigrate](https://www.npmjs.com/package/sqlmigrate) provides hash checks for migration files (disclaimer: I am the author of this one)

**Comment by pleerock** [comment]
Sounds good to me. We can support both features in our migrations, we just need to make sure its not a breaking for exist users. So, feel free to contribute 👍

**Comment by jeremija** [comment]
Thanks, will try to come up with something over the next few weeks!

**Comment by qumberrizvi** [comment]
Being able to hash the previous migrations would be a great feature to ensure integrity. As aforementioned, many DB migration tools, such as Liquibase and Flyway, provide this functionality out of the box.

**Comment by alexrosenfeld10** [comment]
I'd be interested in this feature as well. @jeremija have you started work on it?

**Comment by afehid** [comment]
I'd like to take this with a non-breaking first slice. Design:

**Columns** (nullable, added automatically if the table already exists):
- `executedAt` — timestamp when the migration row was inserted (useful for production debugging)
- `checksum` — SHA-256 of the loaded migration's `up`/`down` function source + class name

**Compatibility**
- Existing tables keep working: new columns are nullable, old rows stay `NULL`
- Checksum **verification is opt-in** via `migrationsChecksumCheck: true` on DataSource options (default `false`) so current projects do not fail if a past migration was edited
- When enabled, only rows that already have a checksum are compared; `NULL` checksums are skipped
- Mismatch throws a clear error; users who intentionally edit an old migration can disable the check or fake-run after fixing

**Out of scope for this PR**
- Arbitrary extra metadata columns
- Hashing original `.ts` files (migrations are often loaded as compiled classes)

I'll open a PR against this issue. Happy to adjust if you'd rather ship `executedAt` first and checksum later.

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Add metadata to migrations table`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Add metadata to migrations table`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Add metadata to migrations table`). I'll dig into the root cause and follow up with a PR shortly. (claiming via @LeonxLJX)
